### PR TITLE
Xor pointers

### PR DESCRIPTION
**Pull request by jeremysinger**
this is the horrible XOR'd pointers code for a doubly-linked list with a single pointer field (superposition of prev and next pointers). 

I don't think this code will work on CHERI ... this is a vanilla C version for now.

### CODE REVIEW

File: xor_pointers.c [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+	cell_t *cell = (cell_t *)malloc(sizeof(cell_t));
+	cell->data = payload;
+	/* xor this cell's address into the tail's ptr field */
+	if (tail != NULL)
+		tail->ptr ^= (WORD)cell;
+	/* store tail address in this cell's ptr field
+	 * NOTE most recently allocated cell (i.e. head)
+	 * only stores a single ptr, not an XOR'd pair
+	 */
+	cell->ptr = (WORD)tail;
```
Comment (cristianurlea) [inline xor_pointers.c:34]: If I understand correctly, assuming a non-CHERI system there's no way to tell if this node is the "last" one in a chain, aside from from the implicit state of the system.

On a CHERI system however,  a check on `cell->ptr` being a valid capability pointer reveals  this information "for free".

File: xor_pointers.c [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+	cell_t *head = NULL;
+	cell_t *prev;
+	cell_t *curr;
+	cell_t *next;
+	for (i = 0; i < NUM_CELLS; i++)
+	{
+		/* allocate a cell */
+		head = alloc_cell(i, head);
+	}
+	/* now we have 100 cells in a doubly-linked list */
```
Comment (cristianurlea) [inline xor_pointers.c:50]: Optional: /s/100/NUM_CELLS

**Comment by ltratt** [comment]
@jeremysinger Is this PR stalled or not quite right for inclusion or ...?

**Comment by ltratt** [comment]
@jeremysinger I propose we close this PR unless you object. It's hanging around in my "assigned" list and I've almost stopped noticing it which is always a sign of a PR that's stuck in one way or another :)

File: xor_pointers.c (hunk 2 above)
Comment (jeremysinger) [reply xor_pointers.c]: fixed in commit bf967e6

**Comment by jeremysinger** [comment]
@ltratt @cristianurlea I hope I have added appropriate commentary for this example - please can we merge now?

**Comment by cristianurlea** [comment]
> @ltratt @cristianurlea I hope I have added appropriate commentary for this example - please can we merge now?

It does everything listed on the tin. I'm all for it.

**Comment by ltratt** [comment]
Hang on, this isn't quite ready for merging! [Even ignoring the bors / buildbot problem.] I need to take a look.

**Comment by ltratt** [comment]
Ok, I've done a review, and I'm happy (and so is @cristianurlea). @jeremysinger Please squash as you see fit.

**Comment by jeremysinger** [comment]
thanks @ltratt - time for some more squashing adventures for me!

**Comment by ltratt** [comment]
This one should be easier I hope :)

**Comment by alex4o** [comment]
Ah, I am really sry for this. I saw Cris's bors command, but the rest of the comments hadn't loaded yet, and I just repeated it.

**Comment by cristianurlea** [comment]
Strange, I thought mine had timed out ?

**Comment by alex4o** [comment]
Yes it timed out, this is why I decided to be helpful and rerun it.

**Comment by ltratt** [comment]
This shouldn't have been merged, but oh well. Let's be more careful next time: we need to make sure that all reviewers are happy before someone triggers bors.

**Comment by cristianurlea** [comment]
Mea culpa. I checked the reviewers list in the top-right section of this page and concluded I was the only one holding this up.

**Pull request merged**